op-devstack/sysgo/l2_cl_supernode.go

```diff
@@ -0,0 +1,327 @@
+	"github.com/ethereum-optimism/optimism/op-service/oppprof"
+	oprpc "github.com/ethereum-optimism/optimism/op-service/rpc"
+	"github.com/ethereum-optimism/optimism/op-service/sources"
+	sncfg "github.com/ethereum-optimism/optimism/op-supernode/config"
+	"github.com/ethereum-optimism/optimism/op-supernode/supernode"
+	sntypes "github.com/ethereum-optimism/optimism/op-supernode/supernode/types"
+)
+
+// SuperNode implements a devstack L2CL node backed by op-supernode, configured for a single chain.
+type SuperNode struct {
+	id     stack.L2CLNodeID
+	p      devtest.P
+	logger log.Logger
+	el     *stack.L2ELNodeID
+
+	// Managed by aggregator
+	svc    *supernode.Supernode
+	cancel context.CancelFunc
+
+	userRPC          string
+	interopEndpoint  string
+	interopJwtSecret eth.Bytes32
+
+	// prepared configs
+	snCfg  *sncfg.CLIConfig
+	vnCfgs map[sntypes.ChainID]*opnodecfg.Config
+}
+
+var _ L2CLNode = (*SuperNode)(nil)
+
+// aggregator maintains a single shared supernode process across multiple chains within a test process.
+var aggregator = &snAggregator{}
+
+type snAggregator struct {
+	mu     sync.Mutex
+	port   int
+	svc    *supernode.Supernode
+	snCfg  *sncfg.CLIConfig
+	vnCfgs map[sntypes.ChainID]*opnodecfg.Config
+	nodes  []*SuperNode
+	logger log.Logger
+}
+
+func (a *snAggregator) addChain(n *SuperNode, p devtest.P, baseCfg *sncfg.CLIConfig, chainID sntypes.ChainID, nodeCfg *opnodecfg.Config, logger log.Logger) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	a.logger = logger
+	if a.vnCfgs == nil {
+		a.vnCfgs = make(map[sntypes.ChainID]*opnodecfg.Config)
+	}
+	a.vnCfgs[chainID] = nodeCfg
+	a.nodes = append(a.nodes, n)
+	if a.snCfg == nil {
+		cfg := *baseCfg
+		cfg.Chains = nil
+		// Reserve a shared port up-front so dependents can form URLs
+		if a.port == 0 {
+			a.port = findFreePort(p)
+		}
+		cfg.RPCConfig.ListenAddr = "127.0.0.1"
+		cfg.RPCConfig.ListenPort = a.port
+		cfg.RPCConfig.EnableAdmin = true
+		a.snCfg = &cfg
+	}
+	// Provide a stable userRPC immediately so dependents (e.g. batcher) can dial
+	n.userRPC = fmt.Sprintf("http://127.0.0.1:%d/%d", a.port, uint64(chainID))
+
+	// If we have multiple chains configured and the service is not yet started,
+	// start the supernode now so downstream services (batcher/proposer) can dial it.
+	if a.svc == nil && len(a.vnCfgs) >= 2 {
+		chains := make([]uint64, 0, len(a.vnCfgs))
+		for cid := range a.vnCfgs {
+			chains = append(chains, uint64(cid))
+		}
+		a.snCfg.Chains = chains
+		ctx, cancel := context.WithCancel(p.Ctx())
+		stopWithCause := func(err error) { cancel() }
+		sn, err := supernode.New(ctx, a.logger, "dev", stopWithCause, a.snCfg, a.vnCfgs)
+		p.Require().NoError(err, "failed to create supernode")
+		a.svc = sn
+		go func() { _ = sn.Start(ctx) }()
+
+		// Wait for each configured chain RPC to expose optimism_rollupConfig
+		for cid := range a.vnCfgs {
+			url := fmt.Sprintf("http://127.0.0.1:%d/%d", a.port, uint64(cid))
+			waitRollupReady(p, logger, url)
+		}
```

Comment:
nit: would this wait-loop be better placed after/outside the start-loop to expedite the boot process? In other words, the wait for the n-th SuperNode RPC will delay the execution of [(n+1)_Supernode.Start()